kv: ship high-resolution tscache summaries during lease transfers and range merges

nvb opened the pull request:
Fixes #61986.
Fixes #117486.
Unblocks #118000.

This commit uses the new `tscache.Cache.Serialize` method introduced in #118299 to ship high-resolution summaries of the timestamp cache during lease transfers and ranges merges. In doing so, it eliminates the loss of precision that occurs in an incoming leaseholder's timestamp cache when it receives a lease transfer or range merge.

This loss of precision was a source of transaction retries for three reasons:
1. txn tombstone marker keys would have their timestamp advanced, leading to TransactionAbortedError with the `ABORT_REASON_NEW_LEASE_PREVENTS_TXN` reason.
2. txn push marker keys would have their timestamp advanced, leading to transactions having their commit timestamp pushed, which could lead to TransactionRetryError with the `RETRY_SERIALIZABLE` reason.
3. global keys would have their timestamp advanced as if they had been read, also leading to transactions having their commit timestamp pushed if they wrote to those keys, which could also lead to TransactionRetryError with the `RETRY_SERIALIZABLE` reason.

The first issue here is the most severe, both because it can not be refreshed away and because it affects transactions of all isolation levels.

This commit introduces two new cluster settings to control the maximum size of these timestamp cache read summaries:
- `kv.lease_transfer_read_summary.local_budget`
- `kv.lease_transfer_read_summary.global_budget`

It configures the local keyspace budget to 4MB and the global keyspace budget to 0B. This default configuration should be sufficient to eliminate the first two sources of retries described above. The third has not been observed as a serious issue in practice, so we configure the global budget to 0 so that we can hit a serialization fast-path.

Release note (ops change): Transaction replay protection state is now passed between the outgoing and incoming leaseholder for a range during a lease transfer. This avoids cases where lease transfers can cause transactions to throw TransactionAbortedError(ABORT_REASON_NEW_LEASE_PREVENTS_TXN) errors.

nvb committed:
kv: ship high-resolution tscache summaries during lease transfers and range merges

Fixes #61986.
Fixes #117486.
Unblocks #118000.

This commit uses the new `tscache.Cache.Serialize` method introduced in #118299 to
ship high-resolution summaries of the timestamp cache during lease transfers and
ranges merges. In doing so, it eliminates the loss of precision that occurs in an
incoming leaseholder's timestamp cache when it receives a lease transfer or range
merge.

This loss of precision was a source of transaction retries for three reasons:
1. txn tombstone marker keys would have their timestamp advanced, leading to
   TransactionAbortedError with the `ABORT_REASON_NEW_LEASE_PREVENTS_TXN`
   reason.
2. txn push marker keys would have their timestamp advanced, leading to
   transactions having their commit timestamp pushed, which could lead to
   TransactionRetryError with the `RETRY_SERIALIZABLE` reason.
3. global keys would have their timestamp advanced as if they had been read,
   also leading to transactions having their commit timestamp pushed if they
   wrote to those keys, which could also lead to TransactionRetryError with the
   `RETRY_SERIALIZABLE` reason.

The first issue here is the most severe, both because it can not be refreshed
away and because it affects transactions of all isolation levels.

This commit introduces two new cluster settings to control the maximum size of
these timestamp cache read summaries:
- `kv.lease_transfer_read_summary.local_budget`
- `kv.lease_transfer_read_summary.global_budget`

It configures the local keyspace budget to 4MB and the global keyspace budget to
0B. This default configuration should be sufficient to eliminate the first two
sources of retries described above. The third has not been observed as a serious
issue in practice, so we configure the global budget to 0 so that we can hit a
serialization fast-path.

Release note (ops change): Transaction replay protection state is now
passed between the outgoing and incoming leaseholder for a range during
a lease transfer. This avoids cases where lease transfers can cause
transactions to throw TransactionAbortedError(ABORT_REASON_NEW_LEASE_PREVENTS_TXN) errors.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118300)
<!-- Reviewable:end -->

nvb committed:
kv: maximally compress persisted prior read summaries

These are rarely read, so it is not worth the extra space to store them
with high precision. Instead, maximally compress them before persisting
them.

Epic: None
Release note: None

arulajmani reviewed (approved):
Exciting to see this land! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-:-Np_byiLEL9tryB7X4Sp:b-xz7486)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_XfLx5SfPnNThEMNA):*
Glad we're no longer imprecise! 
___
*[`-- commits` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_bp7S2YKy6D69lx59):*
They're only read during node failure, right? Or are there other cases I'm missing?
___
*[`pkg/kv/kvserver/replica_tscache.go` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_Yg5f0AUXC1qT8hPX:-Np_Yg5f0AUXC1qT8hPY:b-3ksry0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6700e7eb18ea93dde77cc9629e1892f7a128f57a/pkg/kv/kvserver/replica_tscache.go#L39)):*
> ```Go
> 		"budget will result in loss of precision.",
> 	4<<20, /* 4 MB */
> )
> ```

Did we not want to make this setting public?
___
*[`pkg/kv/kvserver/replica_tscache.go` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_Z-kX3U_PpHJVVMWb:-Np_Z-kX3U_PpHJVVMWc:bt1nio4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6700e7eb18ea93dde77cc9629e1892f7a128f57a/pkg/kv/kvserver/replica_tscache.go#L606)):*
> ```Go
> 			lease, _ /* nextLease */ := r.GetLease()
> 			// Recognize the case where a lease started recently. Lease transfers can
> 			// bump the ts cache low water mark. However, in newer versions of the
> ```

nit: s/newer versions/versions >= 24.1/g
___
*[`pkg/kv/kvserver/replica_tscache.go` line 711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_Zf_K4DesGnUZi3c-:-Np_Zf_K4DesGnUZi3c0:b-o6dx71) ([raw file](https://github.com/cockroachdb/cockroach/blob/6700e7eb18ea93dde77cc9629e1892f7a128f57a/pkg/kv/kvserver/replica_tscache.go#L711)):*
> ```Go
> 			// Serialize the key range and then compress to the budget.
> 			seg = tc.Serialize(ctx, start, end)
> 			seg.Compress(budget)
> ```

We could return a boolean from `Compress`, indicating whether we needed to actually compress or not, and use that to increment some metric that keeps track of the number of times we've run into budget constraints. Feel free to do this in a followup though.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 685 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np__ytM0v9eu_aHmDjR:-Np__ytM0v9eu_aHmDjS:b6iml9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3c3d4e3497629b9891a047c1f2db780e0f34a3b/pkg/kv/kvserver/client_lease_test.go#L685)):*
> ```Go
> 
> 	err = txn.Commit(ctx)
> 	require.Regexp(t, `TransactionAbortedError\(ABORT_REASON_NEW_LEASE_PREVENTS_TXN\)`, err)
> ```

Is it worth keeping a version of the test around that does run into an `ABORT_REASON_NEW_LEASE_PREVENTS_TXN`? Maybe for cases where we've overshot the budget?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-:-NpaTW8C7dw5sutxYVu5:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_XfLx5SfPnNThEMNA):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Glad we're no longer imprecise!
</blockquote></details>

Agreed. Imagine now you are a customer and you read the word "imprecise"...
___
*[`-- commits` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_bp7S2YKy6D69lx59):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

They're only read during node failure, right? Or are there other cases I'm missing?
</blockquote></details>

The main place where they're used is actually when a leaseholder determined that it holds the lease through a Raft snapshot.
___
*[`pkg/kv/kvserver/replica_tscache.go` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_Yg5f0AUXC1qT8hPX:-NpaLNQG51tDbAEpzelc:bcom36b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6700e7eb18ea93dde77cc9629e1892f7a128f57a/pkg/kv/kvserver/replica_tscache.go#L39)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did we not want to make this setting public?
</blockquote></details>

Good idea. I made both public.
___
*[`pkg/kv/kvserver/replica_tscache.go` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_Z-kX3U_PpHJVVMWb:-NpaMDMc-eyG6OcY5nli:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6700e7eb18ea93dde77cc9629e1892f7a128f57a/pkg/kv/kvserver/replica_tscache.go#L606)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/newer versions/versions >= 24.1/g
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_tscache.go` line 711 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np_Zf_K4DesGnUZi3c-:-NpaMjuwDjmi8iiWHMwp:bywg84s) ([raw file](https://github.com/cockroachdb/cockroach/blob/6700e7eb18ea93dde77cc9629e1892f7a128f57a/pkg/kv/kvserver/replica_tscache.go#L711)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We could return a boolean from `Compress`, indicating whether we needed to actually compress or not, and use that to increment some metric that keeps track of the number of times we've run into budget constraints. Feel free to do this in a followup though.
</blockquote></details>

Good idea. I added a TODO for now.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 685 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/118300#-Np__ytM0v9eu_aHmDjR:-NpaO9TwDwsR0NiqEPCK:b-70y67c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3c3d4e3497629b9891a047c1f2db780e0f34a3b/pkg/kv/kvserver/client_lease_test.go#L685)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is it worth keeping a version of the test around that does run into an `ABORT_REASON_NEW_LEASE_PREVENTS_TXN`? Maybe for cases where we've overshot the budget?
</blockquote></details>

Nice idea! Done.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r+
